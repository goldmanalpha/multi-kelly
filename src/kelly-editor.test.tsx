import {
  fireEvent,
  render,
  RenderResult,
} from '@testing-library/react';
import KellyEditor, { Props } from './kelly-editor';
import React from 'react';

const testData = () => ({
  title: 'test: Generous Coin Flip',

  scenarioOutcomes: [
    {
      name: 'test: heads',
      probabilityPct: 22,
      expectedReturnPct: 225,
    },
    {
      name: 'test: tails',
      probabilityPct: 78,
      expectedReturnPct: -25,
    },
  ],
});

const titleText = 'Multi Kelly Criterion Calculator';

let lastRenderResult: RenderResult;

const renderIt = (
  useCallback: boolean = false,
  options: Partial<Props> = {}
) =>
  (lastRenderResult = render(
    <KellyEditor
      startScenarioOutcomes={testData().scenarioOutcomes}
      saveCallback={useCallback ? () => {} : undefined}
      {...options}
    />
  ));

describe('kelly editor', () => {
  const simpleDisplay = (useCallback: boolean) => {
    const { getByText, getByDisplayValue } =
      renderIt(useCallback);

    const title = getByText(titleText);

    expect(title).toBeVisible();

    let elementTests = 0;
    testData().scenarioOutcomes.forEach((so) => {
      const ancestorListItem = getByDisplayValue(
        so.name
      ).closest('li');

      Object.values(so).forEach((v) => {
        const element = getByDisplayValue(v);
        expect(element).toBeVisible();

        // on same row:
        expect(element.closest('li')).toBe(
          ancestorListItem
        );
        elementTests++;
      });
    });

    expect(elementTests).toBe(6);
  };
  it('renders / displays submitted scenario -- without save button', () => {
    simpleDisplay(false);

    expect(lastRenderResult.queryByText('Save')).toBeNull();
    expect(lastRenderResult.container).toMatchSnapshot();
  });

  it('renders / displays submitted scenario -- with save button', () => {
    simpleDisplay(true);

    expect(
      lastRenderResult.queryByText('Save')
    ).toBeEnabled();
    expect(lastRenderResult.container).toMatchSnapshot();
  });

  const errorChecks = (
    modifier: string = '',
    useCallback: boolean = false,
    it: (name: string, fn: () => void) => void = global.it
  ) =>
    describe(`errors ${modifier}`, () => {
      it("shows error when percentages don't add to 100", () => {
        const expectedError =
          "Total probability should be 100% but is 99%. Can't calculate. Please update the probabilities.";
        const { getByText, getByDisplayValue } =
          renderIt(useCallback);

        const inputProbability = getByDisplayValue('22');

        fireEvent.change(inputProbability, {
          target: { value: '21' },
        });

        expect(getByText(expectedError)).toBeVisible();
      });

      it('shows error when name is missing', () => {
        const expectedError =
          'Outcome name is required, please enter a value.';

        const { getByText, getByDisplayValue } =
          renderIt(useCallback);

        const name = getByDisplayValue('test: heads');

        fireEvent.change(name, {
          target: { value: '' },
        });

        expect(getByText(expectedError)).toBeVisible();
      });

      it('shows both name and percent errors', () => {
        const expectedError =
          "Total probability should be 100% but is 99%. Can't calculate. Please update the probabilities. Outcome name is required, please enter a value.";
        const { getByDisplayValue, getByText } =
          renderIt(useCallback);

        const name = getByDisplayValue('test: heads');

        fireEvent.change(name, {
          target: { value: '' },
        });
        const inputProbability = getByDisplayValue('22');

        fireEvent.change(inputProbability, {
          target: { value: '21' },
        });

        expect(getByText(expectedError)).toBeVisible();
      });
    });

  errorChecks();

  const getSaveButton = () =>
    lastRenderResult.getByText('Save').closest('button');

  errorChecks(
    'set save button disabled',
    true,
    (name: string, fun: Function) => {
      it(`${name}`, () => {
        fun();

        expect(getSaveButton()).toBeDisabled();
      });
    }
  );

  it('hides the title when requested', () => {
    const { queryByText } = renderIt(true, {
      showHeader: false,
    });

    expect(queryByText(titleText)).toBeNull();
  });

  it('calls the save callback', (done) => {
    renderIt(true, {
      saveCallback: (summary) => {
        expect(summary.scenarioOutcomes).toEqual(
          testData().scenarioOutcomes
        );
        expect(summary.graph.length).toBe(101);

        expect(summary).toEqual(
          expect.objectContaining({
            betPct: 53.3,
            expectedPayoff: 1.0637941932867376,
          })
        );
        done();
      },
    });

    fireEvent.click(getSaveButton()!);
  });
});
