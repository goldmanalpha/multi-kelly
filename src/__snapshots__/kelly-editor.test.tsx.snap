// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kelly editor renders / displays submitted scenario -- with save button 1`] = `
<div>
  <div
    class="kelly-editor-widget lib-styling"
  >
    <h1
      class="MuiTypography-root MuiTypography-h3"
    >
      Multi Kelly Criterion Calculator
    </h1>
    <div
      class="scenarioManager"
    >
      <div
        class="results"
      >
        
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            kelly bet size: 
            53.3
            %
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            expected return:
             
            6.4
            %
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Save
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <ol
      class="outcomes-container"
    >
      <li>
        <div
          class="scenario lib-styling"
        >
          <div
            class="MuiFormControl-root"
          >
            <span
              title="likelihood of occurence from 0 - 100"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                data-shrink="true"
                for="mui-pct-7"
              >
                Prob Pct*
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline percent-input MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                required=""
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  id="mui-pct-7"
                  name="probabilityPct"
                  required=""
                  type="number"
                  value="22"
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    %
                  </p>
                </div>
              </div>
            </span>
          </div>
          <div
            class="MuiFormControl-root"
          >
            <span
              title="Expected Payoff Pct: gain/loss expected for this scenario
100 = doubling/getting back amount bet twice.
0 = no gain/loss -- just return of amount bet
-100 = losing amount bet."
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                data-shrink="true"
              >
                Exp Gain*
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline percent-input MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                required=""
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  name="expectedReturnPct"
                  required=""
                  type="number"
                  value="225"
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    %
                  </p>
                </div>
              </div>
            </span>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root name"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
              data-shrink="true"
            >
              Name
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="name"
                required=""
                type="text"
                value="test: heads"
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root description"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              Description
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="description"
                type="text"
                value=""
              />
            </div>
          </div>
          <span
            class="delete-scenario"
            title="delete outcome"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </span>
        </div>
      </li>
      <li>
        <div
          class="scenario lib-styling"
        >
          <div
            class="MuiFormControl-root"
          >
            <span
              title="likelihood of occurence from 0 - 100"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                data-shrink="true"
                for="mui-pct-8"
              >
                Prob Pct*
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline percent-input MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                required=""
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  id="mui-pct-8"
                  name="probabilityPct"
                  required=""
                  type="number"
                  value="78"
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    %
                  </p>
                </div>
              </div>
            </span>
          </div>
          <div
            class="MuiFormControl-root"
          >
            <span
              title="Expected Payoff Pct: gain/loss expected for this scenario
100 = doubling/getting back amount bet twice.
0 = no gain/loss -- just return of amount bet
-100 = losing amount bet."
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                data-shrink="true"
              >
                Exp Gain*
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline percent-input MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                required=""
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  name="expectedReturnPct"
                  required=""
                  type="number"
                  value="-25"
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    %
                  </p>
                </div>
              </div>
            </span>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root name"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
              data-shrink="true"
            >
              Name
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="name"
                required=""
                type="text"
                value="test: tails"
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root description"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              Description
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="description"
                type="text"
                value=""
              />
            </div>
          </div>
          <span
            class="delete-scenario"
            title="delete outcome"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </span>
        </div>
      </li>
    </ol>
    <span
      class="add-outcome"
      title="add outcome"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`kelly editor renders / displays submitted scenario -- without save button 1`] = `
<div>
  <div
    class="kelly-editor-widget lib-styling"
  >
    <h1
      class="MuiTypography-root MuiTypography-h3"
    >
      Multi Kelly Criterion Calculator
    </h1>
    <div
      class="scenarioManager"
    >
      <div
        class="results"
      >
        
        <div>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            kelly bet size: 
            53.3
            %
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            expected return:
             
            6.4
            %
          </p>
        </div>
      </div>
    </div>
    <ol
      class="outcomes-container"
    >
      <li>
        <div
          class="scenario lib-styling"
        >
          <div
            class="MuiFormControl-root"
          >
            <span
              title="likelihood of occurence from 0 - 100"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                data-shrink="true"
                for="mui-pct-1"
              >
                Prob Pct*
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline percent-input MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                required=""
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  id="mui-pct-1"
                  name="probabilityPct"
                  required=""
                  type="number"
                  value="22"
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    %
                  </p>
                </div>
              </div>
            </span>
          </div>
          <div
            class="MuiFormControl-root"
          >
            <span
              title="Expected Payoff Pct: gain/loss expected for this scenario
100 = doubling/getting back amount bet twice.
0 = no gain/loss -- just return of amount bet
-100 = losing amount bet."
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                data-shrink="true"
              >
                Exp Gain*
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline percent-input MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                required=""
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  name="expectedReturnPct"
                  required=""
                  type="number"
                  value="225"
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    %
                  </p>
                </div>
              </div>
            </span>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root name"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
              data-shrink="true"
            >
              Name
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="name"
                required=""
                type="text"
                value="test: heads"
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root description"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              Description
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="description"
                type="text"
                value=""
              />
            </div>
          </div>
          <span
            class="delete-scenario"
            title="delete outcome"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </span>
        </div>
      </li>
      <li>
        <div
          class="scenario lib-styling"
        >
          <div
            class="MuiFormControl-root"
          >
            <span
              title="likelihood of occurence from 0 - 100"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                data-shrink="true"
                for="mui-pct-2"
              >
                Prob Pct*
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline percent-input MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                required=""
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  id="mui-pct-2"
                  name="probabilityPct"
                  required=""
                  type="number"
                  value="78"
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    %
                  </p>
                </div>
              </div>
            </span>
          </div>
          <div
            class="MuiFormControl-root"
          >
            <span
              title="Expected Payoff Pct: gain/loss expected for this scenario
100 = doubling/getting back amount bet twice.
0 = no gain/loss -- just return of amount bet
-100 = losing amount bet."
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                data-shrink="true"
              >
                Exp Gain*
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline percent-input MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
                required=""
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
                  name="expectedReturnPct"
                  required=""
                  type="number"
                  value="-25"
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
                  >
                    %
                  </p>
                </div>
              </div>
            </span>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root name"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
              data-shrink="true"
            >
              Name
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="name"
                required=""
                type="text"
                value="test: tails"
              />
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root description"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
              data-shrink="false"
            >
              Description
            </label>
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                name="description"
                type="text"
                value=""
              />
            </div>
          </div>
          <span
            class="delete-scenario"
            title="delete outcome"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
          </span>
        </div>
      </li>
    </ol>
    <span
      class="add-outcome"
      title="add outcome"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
        />
      </svg>
    </span>
  </div>
</div>
`;
